### PR TITLE
Addressing issue #7434: Adding kwarg to Figure.clf() to refetch rcParams and update existing object

### LINKED ISSUE #7434: consult rcParams for Figures on `fig.clf`
version: 2.0.0b1+1773.g3590ce2
installed from src, OSX, gtk backend

if I have in rcParams:
figure.subplot.top=0.9

and I do:
```
figure()
clf()
plot([],[])
draw()
show()
```

then, the top margin is respected, but if I already have a window with figure already open and I do

```
style.use('x') # where x.mplstyle has figure.subplot: 0.5
clf()
plot([],[])
draw()
```

then then top margin is not altered from 0.9 (like I might expect when changing the style).
I have to close the window and create a new one for this style change to have an effect. This is not expected behavior.

### PR DESCRIPTION
**Pull request by nasgold**
## PR Summary
Adding `refetch_rcparams` kwarg to `Figure.clf()` that defaults to `False` to maintain existing, desired behavior. If set to `True`, however, the `Figure.clf()` call will re-consult `rcParams` and update the already extant object. Some (myself included) might find the default behavior counterintuitive, and adding in this kwarg makes behavior either way more explicit.

If the PR is accepted I will document the API change and make an entry in the "next_whats_new" file.

closes #7434  [@tacaswell edited to add GH xrefs]
## PR Checklist

- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [x] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

### CODE REVIEW

File: lib/matplotlib/tests/test_figure.py [hunk 1]
```diff
@@ -119,7 +120,63 @@ def test_clf_keyword():
 
     fig2, ax2 = plt.subplots(2, 1, num=1, clear=True)
     assert fig0 is fig2
-    assert [t.get_text() for t in fig2.texts] == []
```
Comment (jklymak) [inline lib/matplotlib/tests/test_figure.py:122]: I think you dropped this line by accident?

File: lib/matplotlib/tests/test_figure.py [hunk 2]
```diff
@@ -3,7 +3,7 @@
 from pathlib import Path
 import platform
 from threading import Timer
-from types import SimpleNamespace
+from types import BuiltinFunctionType, SimpleNamespace
```
Comment (jklymak) [inline lib/matplotlib/tests/test_figure.py:6]: I don't see where you are using this:
```suggestion
from types import SimpleNamespace
```

File: lib/matplotlib/figure.py [hunk 3]
```diff
@@ -2770,11 +2778,62 @@ def clf(self, keep_observers=False):
...
+            # then utilize local vars to update Figure attributes...
+            # figsize adjustments
+            if not (
+                np.isfinite(figsize).all()
+                or (np.array(figsize) < 0).any()
+            ):
+                raise ValueError(
+                    f"figure size must be positive finite not {figsize}"
+                )
+            self.bbox_inches = Bbox.from_bounds(0, 0, *figsize)
```
Comment (jklymak) [inline lib/matplotlib/figure.py:2799]: OK, this is all copied from `__init__` which is great, except its usually better not to copy (suppose we want to add a new rcParam, now we need to add it in two places).  Can this be cleanly refactored into a helper that is called by both `__init__` and `clf`?  Thanks!

**Comment by tacaswell** [comment]
Well, this is what I suggested in #7434!   If we are going to do this, I still think this is the way to do it, but do we think this is a good idea?

@nasgold In your comment you seem rather down on this feature!  Given that this has been sitting from ~2016 with no additional agitation, maybe this is something that we do not actually need? 

This PR raises a question of if `ax.cla` should also get a similar flag?  It also jumps into a discussion we have been slowly starting to have of what _does_ clearing a figure/axes even mean?  From looking through the code it looks like the people implementing the code have not had a unified opinion on that (based on how throughly different parts of the code "clear" things).

-----

I agree with all of @jklymak 's line comments   I also suspect you will need to run https://github.com/matplotlib/matplotlib/blob/master/tools/boilerplate.py to update the signature and docstring of the auto-generated pyplot functions.

-----

No matter where this lands, thank you for you work on this @nasgold !  Even if we ultimately close this PR (and the other draft PR to address #7434)  and #7434 as "wont-fix", re-surfacing this for discussion is valuable.

**Comment by tacaswell** [comment]
This PR is affected by a re-writing of our history to remove a large number of accidentally committed files [see discourse](https://discourse.matplotlib.org/t/minor-re-writing-of-history-and-moving-from-master-to-main-for-default-branch/22354/6) for details.

To recover this PR it will need be rebased onto the new default branch (main).  There are several ways to accomplish this, but we recommend (assuming that you call the matplotlib/matplotlib remote `"upstream"`

```bash
git remote update
git checkout main
git merge --ff-only upstream/main
git checkout YOUR_BRANCH
git rebase --onto=main upstream/old_master
# git rebase -i main # if you prefer
git push --force-with-lease   # assuming you are tracking your branch
```

If you do not feel comfortable doing this or need any help please reach out to any of the Matplotlib developers.  We can either help you with the process or do it for you.

Thank you for your contributions to Matplotlib and sorry for the inconvenience.

**Comment by tacaswell** [comment]
I am going to close this PR because it has been quiet for more than a year,  still needs the rebase from our history-rewrite issue, and because we still have not decided what "should" happen in these cases.

Thank you for your work on this @nasgold even if we did not ultimately merge it.  I hope we will hear from you again.

**Pull request closed without merging**